_posts/2020-09-07-spring-boot-health-check.md

```diff
@@ -0,0 +1,365 @@
+         }
+      },    
+      "diskSpace": {
+         "status": "UP",
+         "details": {
+            "total": 250685575168,
+            "free": 12073996288,
+            "threshold": 10485760,
+            "exists": true
+         }
+      },
+      "ping": {
+         "status": "UP"
+      }
+   }
+}
+```
+As we can see from the output the health status is composed of status contributed by multiple components called 'Health Indicator' in the Actuator vocabulary. In our case, our health status is composed of health indicators of diskspace and database.
+
+### Construct of a Health Indicator
+Spring Boot Actuator comes with several predefined health indicators like DataSourceHealthIndicator, MongoHealthIndicator, RedisHealthIndicator, CassandraHealthIndicator, etc. Each of them is a class with `Component` annotation that implements the HealthIndicator interface and checks the health of that component. 
+
+By default, all standard components (such as DataSource) implement this interface. The HealthCheck provided by a DataSource is to make a connection to a database and perform a simple query, such as select 1 from dual for instance, against an Oracle DataBase.
+
+These health indicators are used as part of the health check-up process. If our application uses Redis, the RedisHealthIndicator is used for the health check. The MongoHealthIndicator is used for health check if our application uses MongoDB, so on. 
+
+### Aggregating Health Indicators
+The aggregation is done by an implementation of StatusHealthAggregator which aggregates the statuses from each health indicator into a single overall status.
+
+Spring Boot auto-configures an instance of SimpleHealthAggregator. We can provide our implementation of StatusHealthAggregator to supersede the default behavior. 
+
+We can also disable a particular health indicator using application properties :
+```
+management.health.mongo.enabled=false
+```
+
+## Checking the Health of APIs with Custom Health Indicators
+Predefined Health Indicators do not cover all use cases of Health Check. For example, if our API is dependent on any external service, we might like to know if the external service is available. 
+
+In our example, we are using an external service for shortening the URLs. We will monitor the availability of this service by building our health indicator. 
+
+Creating custom health indicators is done in two steps:
+1. Implement the HealthIndicator interface and override the health method.
+2. Register the health indicator class as Spring Bean by adding the `component` annotation.
+
+Our custom Health Indicator for UrlShortener Service looks like this:
+
+```java
+@Component
+@Slf4j
+public class UrlShortenerServiceHealthIndicator implements HealthIndicator, HealthContributor {
+
+  private static final String URL = "https://cleanuri.com/api/v1/shorten";
+
+  @Override
+  public Health health() {
+    // check if url shortener service url is reachable
+        try {
```

Comment:
This code snippet looks a bit too complicated. This can be done with a single try-catch block instead of two nested ones, no?